mm/kfence/kfence-test.c

```diff
@@ -425,6 +425,32 @@ static void test_free_bulk(struct kunit *test)
 	}
 }
 
+/* Test init-on-free works. */
+static void test_init_on_free(struct kunit *test)
+{
+	const int size = 32;
+	struct expect_report expect = {
+		.type = KFENCE_ERROR_UAF,
+		.fn = test_init_on_free,
+	};
+	int i;
+
+	if (!IS_ENABLED(CONFIG_INIT_ON_FREE_DEFAULT_ON))
+		return;
+	/* Assume it hasn't been disabled on command line. */
+
+	setup_test_cache(test, size, NULL);
+	expect.addr = test_alloc(test, size, GFP_KERNEL, ALLOCATE_ANY);
+	for (i = 0; i < size; i++)
+		expect.addr[i] = i + 1;
+	test_free(expect.addr);
+
+	for (i = 0; i < size; i++)
+		KUNIT_EXPECT_EQ(test, expect.addr[i], (char)0);
```

Comment:
While this is currently true, I don't think it's safe to assume that we'll keep this page unprotected forever.

Reply:
Then we'll just unprotect it again. (The test is checking that we encountered an UAF.) The data must always be accessible and the page protection does not prevent accesses to the data. 

The only way this will fail if the page was reallocated and somebody else wrote to it, however the likelihood of this page having been reallocated to some other KFENCE alloc is near impossible.

I'll try to add a comment.

Reply:
Oh no I think I know what you mean. I'll fix the test.